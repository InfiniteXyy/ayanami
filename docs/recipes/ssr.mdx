---
name: SSR
route: /en/recipes/ssr
menu: Recipes
---

# SSR

## `SSREffect`
`SSREffect` is a replacement to `Effect` when you want do **Server Side Rendering**.

## `runSSREffects`
