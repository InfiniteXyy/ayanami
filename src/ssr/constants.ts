export const SSRSymbol = Symbol('__AyanamiSSR__')
export const DEFAULT_SCOPE_NAME = '_$DEFAULT_SCOPE$_'
export const GLOBAL_KEY = Symbol.for('__GLOBAL_MODULE_CACHE__')
export const SSR_LOADED_KEY = Symbol('SSR_LOADED')
export const ACTION_TO_SKIP_KEY = Symbol('ACTION_TO_SKIP')
