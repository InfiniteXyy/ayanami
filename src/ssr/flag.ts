export const SSREnabled =
  typeof process.env.ENABLE_AYANAMI_SSR !== 'undefined'
    ? process.env.ENABLE_AYANAMI_SSR
    : typeof process !== 'undefined' &&
      process.versions &&
      typeof process.versions.node === 'string'
