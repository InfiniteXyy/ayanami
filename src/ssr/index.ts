export * from './express'
export * from './constants'
export * from './constants'
export * from './ssr-module'
export * from './terminate'
export { emitSSREffects, ModuleMeta } from './run'
export * from './flag'
export * from './ssr-context'
