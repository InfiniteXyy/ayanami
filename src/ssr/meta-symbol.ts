export const SSRSymbol = Symbol('__AyanamiSSR__')
