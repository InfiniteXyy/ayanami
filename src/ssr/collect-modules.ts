import { flattenDeep, uniqWith } from 'lodash'

import { Ayanami } from '../core/ayanami'
import { ConstructorOf } from '../core/types'
import { SSRSymbol } from './meta-symbol'

export type ModuleMeta =
  | ConstructorOf<Ayanami<any>>
  | { module: ConstructorOf<Ayanami<any>>; scope: string }

export const activedModulesSets = new Map<object, Set<ModuleMeta>>()
export const collectModuleCallbacks: ((identity: object) => void)[] = []

export const collectModules = (modules: ModuleMeta[][] | undefined) => {
  if (modules) {
    uniqWith(
      flattenDeep(modules).filter((m) => !!m) as ModuleMeta[],
      (meta1: any, meta2: any) =>
        meta1.scope &&
        meta2.scope &&
        meta1.module &&
        meta2.module &&
        meta1.module === meta2.module &&
        meta1.scope === meta2.scope,
    ).forEach((m) => {
      collectModuleCallbacks.push((identity: object) => {
        const existed = activedModulesSets.get(identity)
        if (existed) {
          existed.add(m)
        } else {
          activedModulesSets.set(identity, new Set([m]))
        }
      })
    })
  }
}

export const collectDynamicModule = (component: any, ...metas: ModuleMeta[]) => {
  if (component[SSRSymbol]) {
    component[SSRSymbol] = component[SSRSymbol].concat(metas)
  } else {
    component[SSRSymbol] = metas
  }
}
