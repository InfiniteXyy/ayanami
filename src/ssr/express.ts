import { Request } from 'express'
import { Observable } from 'rxjs'
import { skip } from 'rxjs/operators'

import { SSRSymbol } from './meta-symbol'
import { SSREnabled } from './flag'
import { Effect } from '../core/decorators'

export const SKIP_SYMBOL = Symbol('skip')

function addDecorator(target: any, method: any, middleware: any) {
  const existedMetas = Reflect.getMetadata(SSRSymbol, target)
  const meta = { action: method, middleware }
  if (existedMetas) {
    existedMetas.push(meta)
  } else {
    Reflect.defineMetadata(SSRSymbol, [meta], target)
  }
}

export function SSR<T, Payload>(
  middleware?: (
    req: Request,
    skip: () => typeof SKIP_SYMBOL,
  ) => Payload | Promise<Payload> | typeof SKIP_SYMBOL,
) {
  return (target: T, method: string, descriptor: PropertyDescriptor) => {
    addDecorator(target, method, middleware)
    if (!SSREnabled) {
      const originalValue = descriptor.value
      descriptor.value = function(
        this: any,
        action$: Observable<Payload>,
        state$?: Observable<any>,
      ) {
        if (Reflect.getMetadata(this.ssrLoadKey, this)) {
          return originalValue.call(this, action$.pipe(skip(1)), state$)
        }
        return originalValue.call(this, action$, state$)
      }
    }
    return Effect()(target, method, descriptor)
  }
}

export function SSRServerOnly<T, Payload>(
  middleware?: (
    req: Request,
    skip: () => typeof SKIP_SYMBOL,
  ) => Payload | Promise<Payload> | typeof SKIP_SYMBOL,
) {
  return (target: T, method: string, descriptor: PropertyDescriptor) => {
    addDecorator(target, method, middleware)
    return Effect()(target, method, descriptor)
  }
}
