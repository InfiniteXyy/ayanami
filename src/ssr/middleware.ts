import { Request, Response } from 'express'

export interface SSRMeta<Payload = {}> {
  middleware?: (req: Request, res: Response) => Promise<Payload> | Payload
  action: string
}
export const epicsMap = new Map<object, SSRMeta>()
