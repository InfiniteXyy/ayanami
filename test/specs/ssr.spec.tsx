import * as React from 'react'
import { Observable, timer } from 'rxjs'
import { endWith, switchMap, map, mergeMap } from 'rxjs/operators'
import { Draft } from 'immer'
import { renderToString } from 'react-dom/server'
import { create } from 'react-test-renderer'

import {
  SSRModule,
  Ayanami,
  SSREffect,
  EffectAction,
  emitSSREffects,
  TERMINATE_ACTION,
  ImmerReducer,
  useAyanami,
  SSRContext,
  globalKey,
  reqMap,
} from '../../src'
import { DEFAULT_SCOPE_NAME } from '../../src/ssr/constants'

interface CountState {
  count: number
  name: string
}

@SSRModule('CountModel')
class CountModel extends Ayanami<CountState> {
  defaultState = { count: 0, name: '' }

  @ImmerReducer()
  setCount(state: Draft<CountState>, count: number) {
    state.count = count
  }

  @ImmerReducer()
  setName(state: Draft<CountState>, name: string) {
    state.name = name
  }

  @SSREffect()
  getCount(payload$: Observable<void>): Observable<EffectAction> {
    return payload$.pipe(
      switchMap(() =>
        timer(1).pipe(
          map(() => this.getActions().setCount(1)),
          endWith(TERMINATE_ACTION),
        ),
      ),
    )
  }

  @SSREffect({
    payloadGetter: (req, skip) => req.url || skip(),
    skipFirstClientDispatch: false,
  })
  skippedEffect(payload$: Observable<string>): Observable<EffectAction> {
    return payload$.pipe(
      switchMap((name) =>
        timer(1).pipe(
          map(() => this.getActions().setName(name)),
          endWith(TERMINATE_ACTION),
        ),
      ),
    )
  }
}

interface TipState {
  tip: string
}

@SSRModule('TipModel')
class TipModel extends Ayanami<TipState> {
  defaultState = { tip: '' }

  @ImmerReducer()
  setTip(state: Draft<TipState>, tip: string) {
    state.tip = tip
  }

  @SSREffect()
  getTip(payload$: Observable<void>): Observable<EffectAction> {
    return payload$.pipe(
      mergeMap(() =>
        timer(1).pipe(
          map(() => this.getActions().setTip('tip')),
          endWith(TERMINATE_ACTION),
        ),
      ),
    )
  }
}

const Component = () => {
  const [state, actions] = useAyanami(CountModel)

  React.useEffect(() => {
    actions.setName('new name')
  }, [])

  return <span>{state.count}</span>
}

describe('SSR specs:', () => {
  beforeAll(() => {
    // @ts-ignore
    process.env.ENABLE_AYANAMI_SSR = 'true'
    process.env.NODE_ENV = 'production'
  })

  afterAll(() => {
    // @ts-ignore
    process.env.ENABLE_AYANAMI_SSR = 'false'
    delete process.env.NODE_ENV
  })

  it('should throw if module name not given', () => {
    function generateException() {
      // @ts-ignore
      @SSRModule()
      class ErrorModel extends Ayanami<any> {
        defaultState = {}
      }

      return ErrorModel
    }

    expect(generateException).toThrow()
  })

  it('should pass valid module name', () => {
    @SSRModule('1')
    class Model extends Ayanami<any> {
      defaultState = {}
    }

    @SSRModule({ name: '2', providers: [] })
    class Model2 extends Ayanami<any> {
      defaultState = {}
    }

    function generateException1() {
      @SSRModule('1')
      class ErrorModel1 extends Ayanami<any> {
        defaultState = {}
      }

      return ErrorModel1
    }

    function generateException2() {
      @SSRModule({ name: '1', providers: [] })
      class ErrorModel2 extends Ayanami<any> {
        defaultState = {}
      }

      return { ErrorModel2 }
    }

    function generateException3() {
      // @ts-ignore
      @SSRModule()
      class ErrorModel extends Ayanami<any> {
        defaultState = {}
      }

      return ErrorModel
    }

    expect(Model).not.toBe(undefined)
    expect(Model2).not.toBe(undefined)
    expect(generateException1).toThrow()
    expect(generateException2).toThrow()
    expect(generateException3).toThrow()
    process.env.NODE_ENV = 'development'
    expect(generateException1).not.toThrow()
    expect(generateException2).not.toThrow()
  })

  it('should run ssr effects', async () => {
    // @ts-ignore
    const { state, cleanup } = await emitSSREffects({ url: 'name' }, [CountModel])
    const moduleState = state['CountModel'][DEFAULT_SCOPE_NAME]

    expect(moduleState).not.toBe(undefined)
    expect(moduleState.count).toBe(1)
    expect(moduleState.name).toBe('name')
    expect(state).toMatchSnapshot()
    cleanup()
  })

  it('should skip effect if it returns SKIP_SYMBOL', async () => {
    // @ts-ignore
    const { state, cleanup } = await emitSSREffects({}, [CountModel])
    const moduleState = state['CountModel'][DEFAULT_SCOPE_NAME]

    expect(moduleState.name).toBe('')
    expect(state).toMatchSnapshot()
    cleanup()
  })

  it('should return right state in hooks', async () => {
    const req = {}
    // @ts-ignore
    const { cleanup } = await emitSSREffects(req, [CountModel])
    const html = renderToString(
      // @ts-ignore
      <SSRContext.Provider value={req}>
        <Component />
      </SSRContext.Provider>,
    )
    expect(html).toContain('<span>1</span>')
    expect(html).toMatchSnapshot()
    cleanup()
  })

  it('should work with scope', async () => {
    // @ts-ignore
    const { state, cleanup } = await emitSSREffects({}, [
      { module: CountModel, scope: 'scope' },
      CountModel,
    ])
    const moduleState = state['CountModel'][DEFAULT_SCOPE_NAME]
    const scopedModuleState = state['CountModel']['scope']

    expect(scopedModuleState).not.toBe(undefined)
    expect(scopedModuleState).toEqual(moduleState)
    expect(scopedModuleState).not.toBe(moduleState)
    expect(state).toMatchSnapshot()
    cleanup()
  })

  it('should restore state from global', () => {
    process.env.ENABLE_AYANAMI_SSR = 'false'
    // @ts-ignore
    global[globalKey] = {
      CountModel: {
        [DEFAULT_SCOPE_NAME]: {
          count: 1,
          name: '',
        },
      },
    }
    const testRenderer = create(<Component />)
    expect(testRenderer.root.findByType('span').children[0]).toBe('1')
    process.env.ENABLE_AYANAMI_SSR = 'true'
  })

  it('should timeout and clean', async () => {
    try {
      // @ts-ignore
      await emitSSREffects({}, [CountModel], 0)
    } catch (e) {
      expect(e.message).toContain('Terminate timeout')
    }

    expect(reqMap.size).toBe(0)
  })

  it('should support concurrency', async () => {
    return Promise.all([
      // @ts-ignore
      emitSSREffects({ url: 'name1' }, [
        { module: CountModel, scope: 'scope1' },
        { module: TipModel, scope: 'scope1' },
      ]),
      // @ts-ignore
      emitSSREffects({ url: 'name2' }, [
        { module: CountModel, scope: 'scope1' },
        { module: TipModel, scope: 'scope2' },
      ]),
    ]).then(([result1, result2]) => {
      expect(result1.state['CountModel']['scope1'].name).toBe('name1')
      expect(result2.state['CountModel']['scope1'].name).toBe('name2')
      expect({ firstRequest: result1.state, secondRequest: result2.state }).toMatchSnapshot()
      result1.cleanup()
      result2.cleanup()
    })
  })
})
