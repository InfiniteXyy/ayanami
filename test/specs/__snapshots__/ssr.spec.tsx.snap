// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR specs: should return right state in hooks 1`] = `"<span>1</span>"`;

exports[`SSR specs: should run ssr effects 1`] = `
Object {
  "CountModel": Object {
    "__$$AYANAMI_DEFAULT__SCOPE$$__": Object {
      "count": 1,
      "name": "name",
    },
  },
}
`;

exports[`SSR specs: should skip effect if it returns SKIP_SYMBOL 1`] = `
Object {
  "CountModel": Object {
    "__$$AYANAMI_DEFAULT__SCOPE$$__": Object {
      "count": 1,
      "name": "",
    },
  },
}
`;

exports[`SSR specs: should support concurrency 1`] = `
Object {
  "firstRequest": Object {
    "CountModel": Object {
      "scope1": Object {
        "count": 1,
        "name": "name1",
      },
    },
    "TipModel": Object {
      "scope1": Object {
        "tip": "tip",
      },
    },
  },
  "secondRequest": Object {
    "CountModel": Object {
      "scope1": Object {
        "count": 1,
        "name": "name2",
      },
    },
    "TipModel": Object {
      "scope2": Object {
        "tip": "tip",
      },
    },
  },
}
`;

exports[`SSR specs: should work with scope 1`] = `
Object {
  "CountModel": Object {
    "__$$AYANAMI_DEFAULT__SCOPE$$__": Object {
      "count": 1,
      "name": "",
    },
    "scope": Object {
      "count": 1,
      "name": "",
    },
  },
}
`;
